File: src/Cantina.YargSession/LiveState.cs
// SPDX-License-Identifier: LGPL-3.0-or-later

namespace Cantina.YargSession;

/// <summary>
/// Freshness tiers from <c>docs/live-state.md</c>. The datagram arrives at ~90.7 Hz, so
/// silence is meaningful quickly — but a healthy run has shown a 538 ms gap, so demotions
/// are debounced by the tracker rather than reported raw.
/// </summary>
public enum LiveFreshness
{
    /// <summary>Newest datagram younger than 500 ms.</summary>
    Live = 0,

    /// <summary>Younger than 5 s. The client shows the last state, visibly dimmed.</summary>
    Stale = 1,

    /// <summary>Older than 5 s, no datagram yet, or a socket fault.</summary>
    Dead = 2,
}

/// <summary>
/// Named faults per <c>docs/failure-behavior.md</c>. A fault is reported with its name,
/// never as an empty or frozen state.
/// </summary>
public enum SessionFault
{
    None = 0,

    /// <summary>No datagram has ever arrived on this listener.</summary>
    NoDatagrams = 1,

    /// <summary>The stream was live and stopped.</summary>
    StreamDead = 2,

    /// <summary>
    /// Datagrams from more than one endpoint. The stream is a LAN broadcast (D-020), so a
    /// second YARG anywhere reaches this listener; interleaving two games manufactured a
    /// withdrawn finding once, so this is surfaced, never resolved by picking a sender.
    /// </summary>
    MultipleSources = 3,

    /// <summary>
    /// The port could not be bound: another application holds the YARG data port without
    /// <c>SO_REUSEADDR</c> (D-013). Not retried — retrying cannot succeed and presents
    /// as a hang.
    /// </summary>
    PortConflict = 4,
}

/// <summary>How the current song identity became known (docs/live-state.md).</summary>
public enum SongSource
{
    Unknown = 0,

    /// <summary>Latched from <c>currentSong.json</c>.</summary>
    Observed = 1,

    /// <summary>Cued by Barkeep and confirmed by observation.</summary>
    CuedByBarkeep = 2,
}

/// <summary>Latched song identity. Carried through the score screen (D-010).</summary>
public sealed record LatchedSong(string Title, string Artist, string Hash);

/// <summary>
/// The live-state snapshot promised to the client, per <c>docs/live-state.md</c>.
/// Everything here is observation; nothing is inference. A field this type cannot fill
/// honestly is absent from it by design — there is no playback position and no score.
/// </summary>
public sealed record LiveState
{
    public required YargScene Scene { get; init; }

    public required YargPlayState PlayState { get; init; }

    /// <summary>Latched identity, or null when no song has been observed since the last menu dwell.</summary>
    public required LatchedSong? Song { get; init; }

    public required SongSource SongSource { get; init; }

    /// <summary>Arrival time of the newest accepted datagram, or null before the first.</summary>
    public required DateTimeOffset? ReceivedAt { get; init; }

    public required LiveFreshness Freshness { get; init; }

    public required SessionFault Fault { get; init; }

    /// <summary>Distinct sender endpoints observed, for the MultipleSources report.</summary>
    public required IReadOnlyList<string> Senders { get; init; }

    public required long DatagramsAccepted { get; init; }

    public required long DatagramsRejected { get; init; }
}

File: src/cantina-client/src/liveState.ts
// SPDX-License-Identifier: LGPL-3.0-or-later

// Mirrors Barkeep's LiveState projection (docs/live-state.md). The client renders what
// Barkeep observed and never invents: a stale or dead feed is said out loud, a fault is
// shown by name, and nothing here guesses position or progress.

export type LiveScene =
  | 'Unknown'
  | 'Menu'
  | 'Gameplay'
  | 'Score'
  | 'Calibration'
  | 'Practice'

export type LivePlayState = 'NoSong' | 'Playing' | 'Paused'

export type LiveFreshness = 'Live' | 'Stale' | 'Dead'

export type SessionFault =
  | 'None'
  | 'NoDatagrams'
  | 'StreamDead'
  | 'MultipleSources'
  | 'PortConflict'

export type LatchedSong = {
  title: string
  artist: string
  hash: string
  location: string
}

export type LiveState = {
  scene: LiveScene
  playState: LivePlayState
  song: LatchedSong | null
  songSource: 'Unknown' | 'Observed' | 'CuedByBarkeep'
  receivedAt: string | null
  freshness: LiveFreshness
  fault: SessionFault
  senders: string[]
  datagramsAccepted: number
  datagramsRejected: number
}

export type StageCopy = {
  headline: string
  detail: string
  tone: 'live' | 'attention' | 'down'
}

// The iPad's words for each honest state, per docs/failure-behavior.md: what stands in
// the way, and who can fix it.
export function stageCopy(state: LiveState): StageCopy {
  if (state.fault === 'PortConflict') {
    return {
      headline: 'Another app holds the YARG data port',
      detail: 'Barkeep cannot listen until it is closed.',
      tone: 'down',
    }
  }

  if (state.fault === 'MultipleSources') {
    return {
      headline: 'Two YARG instances are broadcasting',
      detail: 'Close one; Cantina will not guess which game is real.',
      tone: 'down',
    }
  }

  if (state.freshness === 'Dead') {
    return {
      headline: 'YARG is not observable',
      detail:
        state.fault === 'NoDatagrams'
          ? 'No data has arrived. Is YARG running with its UDP stream enabled?'
          : 'The stream stopped. YARG may have exited.',
      tone: 'down',
    }
  }

  const dimmed = state.freshness === 'Stale' ? ' (last known — the feed is stale)' : ''

  if (state.scene === 'Gameplay' && state.song) {
    const verb = state.playState === 'Paused' ? 'Paused' : 'Now playing'
    return {
      headline: `${verb}: ${state.song.title}`,
      detail: `${state.song.artist}${dimmed}`,
      tone: state.playState === 'Paused' ? 'attention' : 'live',
    }
  }

  if (state.scene === 'Score') {
    return {
      headline: 'On the score screen',
      detail: `Players decide what happens next${dimmed}`,
      tone: 'attention',
    }
  }

  return {
    headline: 'At the menus',
    detail: `YARG is between songs${dimmed}`,
    tone: 'live',
  }
}
